internal/authentication/ldap_client_factory.go

```diff
@@ -117,162 +121,393 @@ type PooledLDAPClientFactory struct {
+	// prime the pool to test connectivity and run a quick self-check
+	if !f.tryAddPooledClient(f.pool) {
+		err = fmt.Errorf("LDAP pool couldn't acquire initial client")
+	} else if rerr := f.ReadinessCheck(); rerr != nil {
+		err = rerr // failed readiness check
+	} else {
+		go f.poolManager() // transfer responsibility to the pool manager
+		return nil
 	}
 
-	return nil
+	logging.Logger().Debug("LDAP pool initialization failed, aborting...")
+	err = fmt.Errorf("LDAP pool initialization failed: %w", err)
+	f.closed<-err
+	_ = f.Close()
+	return err
 }
 
-// GetClient opens new client using the pool.
+// thread-safe, opens new client using the pool.
 func (f *PooledLDAPClientFactory) GetClient(opts ...LDAPClientFactoryOption) (conn ldap.Client, err error) {
 	if len(opts) != 0 {
 		return getLDAPClient(f.config.Address.String(), f.config.User, f.config.Password, f.config.Timeout, f.dialer, f.tls, f.config.StartTLS, f.opts, opts...)
 	}
 
-	return f.acquire(context.Background())
+	return f.acquire()
 }
 
+// TODO: Clarify why this method is not thread-safe - or does this only apply to the returnd client?
+
 // The new function creates a pool based client. This function is not thread safe.
+// NOTE: this function adheres to config.Timeout, not config.Pooling.Timeout!
 func (f *PooledLDAPClientFactory) new() (pooled *LDAPClientPooled, err error) {
-	var client ldap.Client
-
-	if client, err = getLDAPClient(f.config.Address.String(), f.config.User, f.config.Password, f.config.Timeout, f.dialer, f.tls, f.config.StartTLS, f.opts); err != nil {
-		return nil, fmt.Errorf("error occurred establishing new client for the pool: %w", err)
-	}
+	client, err := getLDAPClient(f.config.Address.String(), f.config.User, f.config.Password, f.config.Timeout, f.dialer, f.tls, f.config.StartTLS, f.opts)
+	if err != nil { return nil, err }
 
 	return &LDAPClientPooled{Client: client}, nil
 }
 
-// ReleaseClient returns a client using the pool or closes it.
+// thread-safe, returns a client using the pool or closes it.
 func (f *PooledLDAPClientFactory) ReleaseClient(client ldap.Client) (err error) {
-	f.mu.Lock()
-
-	defer f.mu.Unlock()
+	logging.Logger().Trace("Releasing LDAP client")
 
-	if f.closing {
+	pooled, ok := client.(*LDAPClientPooled)
+	if !ok {
+		logging.Logger().Debug("LDAP client is not pooled, closing directly")
 		return client.Close()
 	}
 
-	if pool, ok := client.(*LDAPClientPooled); !ok || cap(f.pool) == len(f.pool) {
-		// Prevent extra or non-pool connections from being returned into the pool.
-		return client.Close()
-	} else {
-		f.pool <- pool
+	if !pooled.IsHealthy() || pooled.IsExpired() {
+		logging.Logger().Info("Pooled LDAP client is not healthy or has expired, disposing")
+		return f.disposeClient(pooled)
 	}
 
-	return nil
+	pC := f.pool
+	if pC == nil {
+		logging.Logger().Warning("LDAP pool is not initialized or was closed, disposing client")
```

Comment:
Please remove call to logging.Logger() and initialize a fielded logger like in #10383.